org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java

```diff
@@ -435,6 +437,19 @@ public static void applyEdits(IDocument document, List<? extends TextEdit> edits
 		}
 	}
 
+	public static void applyEditsWithVersionCheck(IDocument document, VersionedResult<List<? extends TextEdit>> versionedEdits) throws BadLocationException {
```

Comment:
I think any call to applyEdits is affected by the concurrency problem, isn't it? If so, should we not change applyEdits to use versionedEdits?

Reply:
Potentially anything that gets a list of edits from the server is vulnerable to the documents having changed by the time they receive and act on the response, yes. I added support for detecting and avoiding this as a general API that existing code can leverage, as you see, but didn't search through all existing uses of `applyEdits` to fix them as well.

As I mention above, I think in practice the risk is lessened in many cases because of the way the functionality is actually exposed to the user - e.g. rename uses a modal dialog, so the user is blocked from making further changes to the documents while the server is computing the rename delta.